services/src/main/scala/cromwell/pipeline/service/UserService.scala

```diff
@@ -61,23 +61,40 @@ object UserService {
         userId: UserId,
         request: PasswordUpdateRequest,
         salt: String = Random.nextLong().toHexString
-      ): Future[Int] =
-        if (request.newPassword == request.repeatPassword) {
-          userRepository.getUserById(userId).flatMap {
-            case Some(user) =>
-              user match {
-                case user
-                    if user.passwordHash == calculatePasswordHash(request.currentPassword.unwrap, user.passwordSalt) => {
-                  val passwordSalt = salt
-                  val passwordHash = calculatePasswordHash(request.newPassword.unwrap, passwordSalt)
-                  userRepository.updatePassword(user.copy(passwordSalt = passwordSalt, passwordHash = passwordHash))
-                }
-                case _ => Future.failed(new RuntimeException("user password differs from entered"))
-              }
-            case None => Future.failed(new RuntimeException("user with this id doesn't exist"))
+      ): Future[Int] = {
+
+        def checkRequestPassword: Future[Unit] =
+          if (request.newPassword == request.repeatPassword) {
+            Future.unit
+          } else {
+            Future.failed(new RuntimeException("new password incorrectly duplicated"))
+          }
+
+        def checkUserPassword(user: User): Future[Unit] =
+          if (user.passwordHash == calculatePasswordHash(request.currentPassword, user.passwordSalt)) {
+            Future.unit
+          } else {
+            Future.failed(new RuntimeException("new password incorrectly duplicated"))
```

Comment:
Shouldn't this exception message be about wrong current password or something like that?

Reply:
Ooops. Nice catch!

Reply:
I'm surprised that all tests pass :) Fixed this error and added a few tests to make sure it won't happen in the future